fairlearn/metrics/metrics_engine.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+
+from . import GroupMetricResult
+
+_MESSAGE_NON_BINARY = "Array {0} contains values other than 0 and 1"
+_MESSAGE_SIZE_MISMATCH = "Array {0} is not the same size as {1}"
+
+
+def metric_by_groups(metric_function, y_true, y_pred, group_data, sample_weight=None):
+    """ Applies a metric to each subgroup of a set of data
+
+    :param metric_function
+    :type Function with signature (y_true, y_pred, sample_weight=None) which returns a scalar
+
+    :param y_true
+    :type Array of actual results (must be 0 or 1)
+
+    :param y_pred
+    :type Array of predicted results (must be 0 or 1)
+
+    :param group_data
+    :type Array indicating the group to which each result belongs
+
+    :param sample_weight
+    :type Array of weights to apply to each result
+    """
+    _check_binary(y_true, "y_true")
+    _check_binary(y_pred, "y_pred")
+    _check_array_sizes(y_true, y_pred, 'y_true', 'y_pred')
```

Comment:
we don't want to specialize to binary, we have plenty of regression based metrics required for the dashboard. for the time being, let's however restrict attention to 1d-array-like y_true and y_pred (will need to relax that once we work with multi-class classification/regression).

Reply:
Removed `_check_binary`